### PR TITLE
Rule S5876: A new session should be created during user authentication

### LINKED ISSUE #2405: Add rule S5876 for JS/TS
## Rule S5876: A new session should be created during user authentication

[S5876](https://jira.sonarsource.com/browse/RSPEC-5876), part of [MMF-2154](https://jira.sonarsource.com/browse/MMF-2154).

### Summary

This rule detects missing regeneration of session ids for Passport.js. It should raise an issue if the success callback of `passport.authenticate` does not call `session.regenerate`.

### Code Samples

https://github.com/SonarSource/security-expected-issues/tree/master/javascript/rules/hotspots/RSPEC-5876

#### Noncompliant Code Example
```
app.post('/login', 
  passport.authenticate('local', { failureRedirect: '/login' }),
  function(req, res) {
    // Sensitive - no session.regenerate after login
    res.redirect('/');
  });
```

#### Compliant Solution
```
app.post('/login', 
  passport.authenticate('local', { failureRedirect: '/login' }),
  function(req, res) {
    let prevSession = req.session;
    req.session.regenerate((err) => {  // Compliant
      Object.assign(req.session, prevSession);
      res.redirect('/');
    });
  });
```

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #2405

### CODE REVIEW

File: eslint-bridge/tests/rules/session-regeneration.test.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import { RuleTester } from 'eslint';
+import { rule } from 'rules/session-regeneration';
+
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2018 } });
+
+ruleTester.run('Using publicly writable directories is security-sensitive', rule, {
```
Comment (guillaume-dequenne-sonarsource) [inline eslint-bridge/tests/rules/session-regeneration.test.ts:26]: _Minor:_ There is a copy paste issue for the name of the test.

File: eslint-bridge/src/rules/session-regeneration.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://jira.sonarsource.com/browse/RSPEC-5678
```
Comment (guillaume-dequenne-sonarsource) [inline eslint-bridge/src/rules/session-regeneration.ts:20]: _Minor:_ The link is not correct, as the rule key is RSPEC-5876.

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM, just a couple minor comments.
As there are no issues on the ruling, it might be interesting to add one in the custom `file-for-rules` one if you feel like it.

File: eslint-bridge/src/rules/session-regeneration.ts (hunk 2 above)
Comment (saberduck) [reply eslint-bridge/src/rules/session-regeneration.ts:20]: don't drink and code!

**Pull request merged**